includes/sanitizers/class-amp-meta-sanitizer.php

```diff
@@ -0,0 +1,333 @@
+	];
+
+	/**
+	 * Reference to the shared XPath object to query the DOM.
+	 *
+	 * @var DOMXPath
+	 */
+	protected $xpath;
+
+	/**
+	 * The document's <head> element.
+	 *
+	 * @var DOMElement
+	 */
+	protected $head;
+
+	/*
+	 * Tags array keys.
+	 */
+	const TAG_CHARSET        = 'charset';
+	const TAG_VIEWPORT       = 'viewport';
+	const TAG_AMP_SCRIPT_SRC = 'amp_script_src';
+	const TAG_OTHER          = 'other';
+
+	/**
+	 * Associative array of DOMElement arrays.
+	 *
+	 * Each key in the root level defines one group of meta tags to process.
+	 *
+	 * @var array $tags {
+	 *     An array of meta tag groupings.
+	 *
+	 *     @type DOMElement[] $charset        Charset meta tag(s).
+	 *     @type DOMElement[] $viewport       Viewport meta tag(s).
+	 *     @type DOMElement[] $amp_script_src <amp-script> source meta tags.
+	 *     @type DOMElement[] $other          Remaining meta tags.
+	 * }
+	 */
+	protected $meta_tags = [
+		self::TAG_CHARSET        => [],
+		self::TAG_VIEWPORT       => [],
+		self::TAG_AMP_SCRIPT_SRC => [],
+		self::TAG_OTHER          => [],
+	];
+
+	/**
+	 * Charset to use for AMP markup.
+	 *
+	 * @var string
+	 */
+	const AMP_CHARSET = 'utf-8';
+
+	/**
+	 * Viewport settings to use for AMP markup.
+	 *
+	 * @var string
+	 */
+	const AMP_VIEWPORT = 'width=device-width';
+
+	/**
+	 * Sanitize.
+	 */
+	public function sanitize() {
+		$this->xpath = new DOMXPath( $this->dom );
+		$this->head  = $this->ensure_head_is_present();
+		$charset     = $this->detect_charset();
+		$elements    = $this->dom->getElementsByTagName( static::$tag );
+
+		// Remove all nodes for easy reordering later on.
+		$elements = array_map(
+			static function ( $element ) {
+				return $element->parentNode->removeChild( $element );
+			},
+			iterator_to_array( $elements, false )
+		);
+
+		foreach ( $elements as $element ) {
+			/**
+			 * Meta tag to process.
+			 *
+			 * @var DOMElement $element
+			 */
+			if ( $element->hasAttribute( 'charset' ) ) {
+				$this->meta_tags[ self::TAG_CHARSET ][] = $element;
+			} elseif ( 'viewport' === $element->getAttribute( 'name' ) ) {
+				$this->meta_tags[ self::TAG_VIEWPORT ][] = $element;
+			} elseif ( 'amp-script-src' === $element->getAttribute( 'name' ) ) {
+				$this->meta_tags[ self::TAG_AMP_SCRIPT_SRC ][] = $element;
+			} else {
+				$this->meta_tags[ self::TAG_OTHER ][] = $element;
+			}
+		}
+
+		$this->ensure_charset_is_present( $charset );
+
+		if ( ! $this->is_correct_charset() ) { // phpcs:ignore Generic.CodeAnalysis.EmptyStatement
+			// @TODO Re-encode the content into UTF-8.
+			// ... sure?
+		}
+
+		$this->ensure_viewport_is_present();
+
+		$this->process_amp_script_meta_tags();
+
+		$this->re_add_meta_tags_in_optimized_order();
+	}
+
+	/**
+	 * Ensure that the <head> element is present in the document.
+	 *
+	 * @return DOMElement The document's <head> element.
+	 */
+	protected function ensure_head_is_present() {
```

Comment:
I wonder if this is, in fact, needed, because now we prevent processing responses when no `<head>` is present:

https://github.com/ampproject/amp-wp/blob/f15760234e6b2b6bae81769f2e80077fcf6d657d/includes/class-amp-theme-support.php#L2052-L2058

Reply:
I think we should treat the sanitizers independently from the `AMP_Theme_Support` processing here, as we want to produce a stand-alone sanitizer library in the long term.

While it is perfectly fine to do early bails in `AMP_Theme_Support` for various reasons as part of the "WP integration" of the plugin, I think that "asserting" the requirements in the sanitizer should still be done nevertheless.